### ISSUE TITLE
update for v0.6.0

### ISSUE THREAD

**Issue by jlapeyre** [jlapeyre/Symata.jl#72]
Lots of changes to make. Some have been made on local branch. Not yet pushed to github.

**Comment by jlapeyre** [comment]
* "method too new"
* rationalize reading ASTs for expressions across Julia versions.

**Comment by goedman** [comment]
Hi John, on "my" branch I've made the updates for abstract   ...  to Julia 0.6's abstract type ... end but I'm stuck on I think a few places. The first one:

julia> Pkg.test("NMfE")
INFO: Testing NMfE
ERROR: LoadError: LoadError: invalid subtyping in definition of UnitRangeInf
Stacktrace:
 [1] include_from_node1(::String) at ./loading.jl:569
 [2] include(::String) at ./sysimg.jl:14
 [3] include_from_node1(::String) at ./loading.jl:569
 [4] include(::String) at ./sysimg.jl:14
 [5] anonymous at ./<missing>:2
while loading /Users/rob/.julia/v0.6/Symata/src/mxpr.jl, in expression starting on line 521
while loading /Users/rob/.julia/v0.6/Symata/src/Symata.jl, in expression starting on line 45

Is that a change that look familiar to you and is maybe on your local branch? 

I know there is at least one more issue that I see when I comment out UnitRangeInf def, but maybe step by step is better.

Regards,
Rob

**Comment by jlapeyre** [comment]
Hi Rob,
Thanks for the help.  I will push my changes today. I don't want us to duplicate effort. I have fixed the  `UnitRangeInf` issue, but many others remain.  I updated all my other packages for `v0.6` fairly easily. But `Symata` is far more difficult. It will take more work to regain the functionality that `Symata` has under `v0.5`.

Cheers, John

**Comment by jlapeyre** [comment]
I pushed a new branch: [branch v0.6-a](https://github.com/jlapeyre/Symata.jl/tree/v0.6-a).  I have not yet tested it with v0.5, but it almost certainly will not work with v0.5. Maintaining compatibility with v0.5 would not be too hard, but would require some work.

* [method too new](https://github.com/JuliaLang/julia/issues/21356). Symata uses the construction `J( julia code )` to run Julia code and return the result. Eg `f = J( x -> x^2 )`, returns a function that can  be called when evaluating a Symata expression. Defining a function this way in the REPL and then using later in the REPL works fine. But, it no longer works in the Symata test suite. I fixed this in many places by using `invokelatest`, which slows execution considerably. In some cases, `invokelatest` is not enough, so I disabled the tests.

* The AST for `a => b` has been changed in v0.6. I made some quick changes to deal with this. But, it needs to be treated comprehensively.

* `UnitRangeInf` problem.  Julia v0.6 is more strict with unit ranges, which caught a bug in Symata. I intended that `1:Inf` be interpreted as an expression. But, it was actually being evaluated. Julia now sees that `1` and `Inf` are of different types and refuses to convert automatically.  So, I spent some time fixing the `@mkapprule` macro. It remains to clean this up and make the calls to `@mkapprule` uniform again.  Also, `UnitRangeInf` should be renamed. 

* `map` no longer returns a container of the same type as the source container. I had relied on this when constructing `Mxpr`. I wrote a new function `mapmargs` so I can control which container is used now and in the future. Using `mapmargs`  fixed many problems. But, map is used in several other places as well, and is avoided in some places. Some remaining bugs may be related to the new `map` behavior; in particular in the `sympy` interface.

* `quadgk` has been moved to a separate package. This brings up the problem of soft dependence again. Julia will have a general solution in the future, but does not now. So I have my own, borrowed from pieces of other solutions. This makes it harder to maintain compatibility with previous Julia versions. Also, soft dependence and loading on demand combined with the  *method too new* thing has broken some features. I have not yet tried to get around this, so I just disabled them.

* Many other bugs have appeared in going from v0.5 to v0.6. I have fixed many of them, but for many others, have not yet identified their source.

**Comment by goedman** [comment]
Thanks a lot John, I was not trying to push you for a 0.6 version, just trying Symata on Julia 0.6 when I saw this issue. I think doing it right is more important than doing it quickly. Switching full time to Julia 0.6 is still quite some time away (a few months?) I guess looking at graphics and some other packages.
At that point backwards compatibility is not an issue for me. In case of Symata keeping one eye on Julia 0.7/1.0 might in the end be a better way forward for Symata.jl.

On the 'soft dependence' issue, if we're talking about the same topic here, I guess everybody is kind of coming up with a solution for that, in my case I've decided to no longer include graphics (Gadfly.jl) and Mamba.jl dependencies in Stan.jl testing and for the (not published yet) PtFEM packages for now I suppress tests depending on Plots.jl, GR.jl and Symata.jl (in PtFEM/NMfE.jl) when testing on Julia > 0.5. But I admit to not taking the same approach to QuadGK in PtFEM/PtFEM.jl.

I will occasionally switch to branch v0.6-a but I've learned the hard way that switching git branches while playing around on multiple REPLs and an editor is a dangerous workflow.

**Comment by jlapeyre** [comment]
Well, I had already made some fixes for v0.6, and it was a good idea to make them available. I am interested in seeing PtFEM when it is released.

There is certainly merit in doing it correctly rather than quickly. But, I think if I hew too closely to that idea with Symata nothing would ever happen. It is a big project and my time is limited. I was hoping to make something usable enough to attract developers, and I have in fact gotten some help. Because I focused on getting a usable system,  many parts of Symata need to be cleaned up or even rewritten.  But, there are many great projects for which the majority of development is done by one or a few people, so I am not surprised by the outcome thus far.

I am still convinced that Julia is (or will be with a stable API) the best choice for writing a language like Symata.
